### PR TITLE
sys/ztimer: ztimer_remove report success

### PR DESCRIPTION
**Pull request by chrysn**
### Contribution description

An application that set a timer and removes it could previously not reliably tell whether the timer fired without keeping track in an atomic (or otherwise synchronized) variable. This introduces a return value to ztimer_remove that allows building applications that do something precisely once (that is, to do whatever cleanup the timeout's callback would do if it had been called.

This is particularly pressing for Rust code, as the language is very picky about closures declared to run exactly once really being called only once. (And while leaking something is generally considered OK there language-wise, on an embedded system the leakage typically means that some mutex is forever locked).

The alternative for the user is to disable interrupts, query ztimer_is_set, and then to remove the timer. That's cumbersome and likely to be done subtly wrong (by not disabling interrupts and creating a race condition).

A second commit encodes an assumption made in the added ztimer_remove documentation in the ztimer_handler documentation.

### Testing procedure

I wouldn't know how to force the corner cases in which this performs visibly different from just clearing and setting, so unless someone has a better idea I'm testing this by running the unit tests and the existing ztimer tests.

Oddly, ztimer_periodic fails on native -- which is a bit odd as there is nothing obvious that can have broken things. Still investigating, and happy to take any clues.

#### API

Returning something where previously void was returned is not usually considered a breaking API change in RIOT; not even the wrappers for Rust care (although technically it's a breaking change there for users of riot-sys, but I'm not even counting these any more).

#### Code size

When built for particle-xenon, code size in tests stays the same except in test_ztimer_msg (which is weird because ztimer_remove is not linked there. Probably just taking different optimization paths now, where larger jumps result from a different sequence of functions in ROM). Generally, the code of tests does differ, but somehow the compiler seems to make the most out of the changes.

#### Performance

No tests were done; unless there is register pressure in ztimer_remove this should take at most a few cycles per ztimer_remove.

### Issues/PRs references

Closes: https://github.com/RIOT-OS/RIOT/issues/16413

Helps resolving: https://github.com/RIOT-OS/RIOT/issues/14390

There was previous discussion there, which I followed where it's not about installing training wheels on ztimers (which may be a good idea but would go into a separate PR).

### CODE REVIEW

File: sys/include/ztimer.h [hunk 1]
```diff
@@ -408,8 +411,13 @@ unsigned ztimer_is_set(const ztimer_clock_t *clock, const ztimer_t *timer);
  *
  * @param[in]   clock       ztimer clock to operate on
  * @param[in]   timer       timer entry to remove
+ *
+ * @return true if the timer was removed (and thus its callback neither was nor
+ *   will be called by ztimer); false if it fired previously or not set on the
+ *   timer at all.
```
Comment (maribu) [inline sys/include/ztimer.h:417]:
```suggestion
 * @retval  true    The timer was removed (and thus its callback neither was nor
 *                  will be called by ztimer).
 * @retval  false   The timer fired previously or is not set on the @p clock
 *                  at all.
```

File: sys/ztimer/core.c [hunk 2]
```diff
@@ -68,18 +68,22 @@ unsigned ztimer_is_set(const ztimer_clock_t *clock, const ztimer_t *timer)
     return res;
 }
 
-void ztimer_remove(ztimer_clock_t *clock, ztimer_t *timer)
+bool ztimer_remove(ztimer_clock_t *clock, ztimer_t *timer)
 {
+    bool was_removed;
```
Comment (maribu) [inline sys/ztimer/core.c:73]:
```suggestion
    bool was_removed = false;
```

File: sys/ztimer/core.c [hunk 3]
```diff
@@ -68,18 +68,22 @@ unsigned ztimer_is_set(const ztimer_clock_t *clock, const ztimer_t *timer)
...
     unsigned state = irq_disable();
 
     if (_is_set(clock, timer)) {
         _ztimer_update_head_offset(clock);
-        _del_entry_from_list(clock, &timer->base);
+        was_removed = _del_entry_from_list(clock, &timer->base);
 
         _ztimer_update(clock);
+    } else {
+        was_removed = false;
```
Comment (maribu) [inline sys/ztimer/core.c:82]:
```suggestion
```

**Review by maribu — COMMENTED** [review]
some comments inline

**Comment by chrysn** [comment]
Thanks, all good changes. Weirdly, on the ztimer_periodic test is now successful (even though the fixups are definitely just cosmetic on the abstract machine). A case of a test scraping by some performance metric that depends on the compiler's ability to optimize, maybe?

**Comment by kaspar030** [comment]
There was one unrelated test failure (pkg_edhoc on esp32).

**Comment by kaspar030** [comment]
please squash!

**Review by kaspar030 — APPROVED** [review]
ACK.

**Pull request merged**